source/codegen/common_helpers.py

```diff
@@ -328,23 +328,147 @@ def filter_parameters_for_grpc_fields(parameters):
-  service_class_prefix = config['service_class_prefix']
-  return {service_class_prefix: attributes}
+    # If the attributes are already in string categories: those are the groups. Return as-is.
+    first_key = next(iter(attributes), None)
+    if isinstance(first_key, str):
+        return [AttributeGroup(name, attributes, config) for name, attributes in attributes.items()]
+
+    # If there's just one level of attributes: use the service_class_prefix as the group.
+    service_class_prefix = config['service_class_prefix']
+    return [AttributeGroup(service_class_prefix, attributes, config)]
 
 
 def strip_prefix(s: str, prefix: str) -> str:
-  return s[len(prefix) :] if s.startswith(prefix) else s
+    return s[len(prefix):] if s.startswith(prefix) else s
 
 
 def strip_suffix(s: str, suffix: str) -> str:
     return s[: -len(suffix)] if s.endswith(suffix) else s
+
+
+def get_grpc_type_name_for_identifier(data_type, config):
+    """
+    Used to create an identifier string based on the grpc_type
+    i.e., double -> Double. repeated double -> DoubleArray.
+    """
+    grpc_type = get_grpc_type(data_type, config)
+    grpc_type = re.sub(r'^(repeated )(\w+)$', r'\2Array', grpc_type)
+    return ensure_pascal_case(grpc_type)
+
+
+def get_grpc_type_from_ivi(type, is_array, driver_name_pascal):
+    add_repeated = is_array
+    if 'ViSession' in type:
+        type = 'nidevice_grpc.Session'
+    if 'ViBoolean' in type:
+        type = 'bool'
+    if 'ViReal64' in type:
+        type = 'double'
+    if 'ViInt32' in type:
+        type = 'sint32'
+    if 'ViConstString' in type:
+        type = 'string'
+    if 'ViString' in type:
+        type = 'string'
+    if 'ViRsrc' in type:
+        type = 'string'
+    if 'ViChar' in type:
+        if is_array:
+            add_repeated = False
+            type = 'string'
+        else:
+            type = 'uint32'
+    if 'ViReal32' in type:
+        type = 'float'
+    if 'ViAttr' in type:
+        type = driver_name_pascal + "Attributes"
+    if 'ViInt8' in type:
+        if is_array:
+            type = "bytes"
+            add_repeated = False
+        else:
+            type = 'uint32'
+    if 'void*' in type:
+        type = 'fixed64'
+    if 'ViInt16' in type:
+        type = 'sint32'
+    if 'ViInt64' in type:
+        type = 'int64'
+    if 'ViUInt16' in type:
+        type = 'uint32'
+    if 'ViUInt32' in type:
+        type = 'uint32'
+    if 'ViUInt64' in type:
+        type = 'uint64'
+    if 'ViUInt8' in type:
+        if is_array:
+            type = "bytes"
+            add_repeated = False
+        else:
+            type = 'uint32'
+    if 'ViStatus' in type:
+        type = 'sint32'
+    if 'ViAddr' in type:
+        type = 'fixed64'
+    if 'int' == type:
+        type = 'sint32'
+    if "[]" in type:
+        type = type.replace("[]", "")
+
+    return "repeated " + type if add_repeated else type
+
+
+def get_grpc_type(data_type, config):
```

Comment:
This moved from metadata_mutation and I changed the implementation significantly. (But it still generates the same)

@gregstoll